pkg/log/logshipper/logshipper.go

```diff
@@ -0,0 +1,124 @@
+package logshipper
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/sendbuffer"
+)
+
+// TODO: consolidate this var, but where?
+var observabilityIngestTokenKey = []byte("observability_ingest_auth_token")
+
+const (
+	truncatedFormatString = "%s[TRUNCATED]"
+)
+
+type LogShipper struct {
+	sender     *authedHttpSender
+	sendBuffer *sendbuffer.SendBuffer
+	logger     log.Logger
+	knapsack   types.Knapsack
+	stopFunc   context.CancelFunc
+}
+
+func New(k types.Knapsack) (*LogShipper, error) {
+	token, _ := k.TokenStore().Get(observabilityIngestTokenKey)
+
+	logEndpoint, err := logEndpoint(k)
+	if err != nil {
+		return nil, err
+	}
+
+	sender := newAuthHttpSender(logEndpoint, string(token))
+
+	sendInterval := time.Minute * 1
+	if k.Debug() {
+		sendInterval = time.Second * 1
+	}
+
+	sendBuffer := sendbuffer.New(sender, sendbuffer.WithSendInterval(sendInterval))
+	logger := log.NewJSONLogger(sendBuffer)
+
+	return &LogShipper{
+		sender:     sender,
+		sendBuffer: sendBuffer,
+		logger:     logger,
+		knapsack:   k,
+	}, nil
+}
+
+func (ls *LogShipper) Ping() {
+	token, _ := ls.knapsack.TokenStore().Get(observabilityIngestTokenKey)
+	ls.sender.authtoken = string(token)
+
+	endpoint, err := logEndpoint(ls.knapsack)
+	if err != nil {
+		ls.logger.Log("msg", "failed to get endpoint", "err", err)
+		return
+	}
+	ls.sender.endpoint = endpoint
+
+	if !ls.knapsack.LogShippingEnabled() {
+		ls.sendBuffer.DeleteAllData()
+	}
+}
+
+func (ls *LogShipper) Run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	ls.stopFunc = cancel
+	return ls.sendBuffer.Run(ctx)
+}
+
+func (ls *LogShipper) Stop(_ error) {
+	ls.stopFunc()
+}
+
+func logEndpoint(k types.Knapsack) (string, error) {
+	endpoint := k.ObservabilityIngestServerURL()
+
+	if !strings.HasPrefix(endpoint, "http") {
+		scheme := "https"
+		if k.DisableObservabilityIngestTLS() {
+			scheme = "http"
+		}
+		endpoint = fmt.Sprintf("%s://%s", scheme, endpoint)
+	}
+
+	parsedUrl, err := url.Parse(endpoint)
+	if err != nil {
+		return "", err
+	}
+
+	return parsedUrl.String(), nil
+}
+
+func (ls *LogShipper) Log(keyvals ...interface{}) error {
+	if !ls.knapsack.LogShippingEnabled() {
+		return nil
+	}
```

Comment:
I cannot remember -- how performant is this? Vs storing a boolean on logshipper and registering it?

Reply:
I think the values are cached, but just to be safe, I'll store the boolean on log shipper